packages/token-generation-readmodel-checker-verifier/src/utils/utils.ts

```diff
@@ -0,0 +1,1254 @@
+  const readModelService = readModelServiceBuilder(readModel);
+  const [resultsA, resultsB, resultsC] = await Promise.all([
+    platformStatesEntries,
+    tokenGenerationStatesEntries,
+    readModelService.getAllReadModelClients(),
+  ]);
+  return zipClientDataById(
+    resultsA,
+    resultsB,
+    resultsC
+  ).reduce<ClientDifferencesResult>((acc, [a, b, c]) => {
+    // TODO: are missing token entries considered errors or not?
+    if (!c || (c && (!a || b?.length === 0))) {
+      // eslint-disable-next-line functional/immutable-data
+      acc.push([
+        a ? PartialPlatformStatesClientEntry.parse(a) : undefined,
+        b && b.length > 0
+          ? PartialTokenStatesClientEntry.array().parse(b)
+          : undefined,
+        c ? PartialClient.parse(c) : undefined,
+      ]);
+      return acc;
+    }
+
+    const { status: isPlatformStatesCorrect, data: platformClientEntryDiff } =
+      validateClientPlatformStates({
+        platformClientEntry: a,
+        client: c,
+      });
+
+    const {
+      status: isTokenGenerationStatesCorrect,
+      data: tokenClientEntryDiff,
+    } = validateClientTokenGenerationStates({
+      tokenEntries: b,
+      client: c,
+    });
+    if (!isPlatformStatesCorrect || !isTokenGenerationStatesCorrect) {
+      // eslint-disable-next-line functional/immutable-data
+      acc.push([
+        platformClientEntryDiff,
+        tokenClientEntryDiff,
+        PartialClient.parse(c),
+      ]);
+    }
+
+    return acc;
+  }, []);
+}
+
+function validateClientPlatformStates({
+  platformClientEntry,
+  client,
+}: {
+  platformClientEntry: PlatformStatesClientEntry | undefined;
+  client: Client;
+}): {
+  status: boolean;
+  data: PartialPlatformStatesClientEntry | undefined;
+} {
+  const status = !platformClientEntry
+    ? true
+    : extractIdFromPlatformStatesPK<ClientId>(platformClientEntry.PK).id ===
+        client.id &&
+      platformClientEntry.clientKind ===
+        clientKindToTokenGenerationStatesClientKind(client.kind) &&
+      platformClientEntry.clientConsumerId === client.consumerId &&
+      platformClientEntry.clientPurposesIds.every((p) =>
+        client.purposes.includes(p)
+      );
+
+  return {
+    status,
+    data:
+      !status && platformClientEntry
+        ? {
+            PK: platformClientEntry.PK,
+            clientKind: platformClientEntry.clientKind,
+            clientConsumerId: platformClientEntry.clientConsumerId,
+            clientPurposesIds: platformClientEntry.clientPurposesIds,
+          }
+        : undefined,
+  };
+}
+
+function validateClientTokenGenerationStates({
+  tokenEntries,
+  client,
+}: {
+  tokenEntries: TokenGenerationStatesGenericEntry[] | undefined;
+  client: Client;
+}): {
+  status: boolean;
+  data: PartialTokenStatesClientEntry[] | undefined;
+} {
+  // TODO: is this correct?
+  if (!tokenEntries || tokenEntries.length === 0) {
+    return {
+      status: true,
+      data: undefined,
+    };
+  }
```

Comment:
Same question as [this](https://github.com/pagopa/interop-be-monorepo/pull/1119/files#r1818828748).